src/game/logic/object/object.cpp

```diff
@@ -2655,51 +2984,468 @@ void Object::Enter_Group(AIGroup *group)
+        m_valueSighting->Reset();
+    }
+}
+
+void Object::Add_Threat()
+{
+    if (m_threatSighting->Is_Invalid()) {
+        if (Get_Controlling_Player() != nullptr) {
+            if (!Get_Status_Bits().Test(OBJECT_STATUS_UNDER_CONSTRUCTION) && !Is_Effectively_Dead()
+                && Get_Shroud_Clearing_Range() > 0.0f) {
+                m_threatSighting->m_where = *Get_Position();
+                m_threatSighting->m_frame = Get_Template()->Get_Threat_Value();
+                m_threatSighting->m_playerIndex = Get_Controlling_Player()->Get_Player_Mask();
+                m_threatSighting->m_radius = Get_Vision_Range();
+                g_thePartitionManager->Do_Threat_Affect(m_threatSighting->m_where.x,
+                    m_threatSighting->m_where.y,
+                    m_threatSighting->m_radius,
+                    m_threatSighting->m_frame,
+                    m_threatSighting->m_playerIndex);
+            }
+        }
+    } else {
+        captainslog_dbgassert(false, "An Object is adding threat, but hasn't removed his previous value.");
+    }
+}
+
+void Object::Remove_Threat()
+{
+    if (!m_threatSighting->Is_Invalid()) {
+        g_thePartitionManager->Undo_Threat_Affect(m_threatSighting->m_where.x,
+            m_threatSighting->m_where.y,
+            m_threatSighting->m_radius,
+            m_threatSighting->m_frame,
+            m_threatSighting->m_playerIndex);
+        m_threatSighting->Reset();
+    }
 }
 
-void Object::Topple(const Coord3D dir, float speed, int options) {}
+void Object::Look()
+{
+    if (m_friendlyLookSighting->Is_Invalid()) {
+        if (Get_Controlling_Player() != nullptr && !Is_Destroyed() && !Is_Effectively_Dead()) {
+            ContainModuleInterface *contain;
+
+            if (Get_Contained_By() != nullptr) {
+                contain = Get_Contained_By()->Get_Contain();
+            } else {
+                contain = nullptr;
+            }
+
+            if (contain == nullptr || contain->Is_Garrisonable()) {
+                float radius = Get_Shroud_Clearing_Range();
+
+                if (radius > 0.0f) {
+                    unsigned short index = 0;
+                    unsigned short index2;
+
+                    if (Is_KindOf(KINDOF_REVEAL_TO_ALL)) {
+                        index2 = -1;
+                    } else {
+                        for (int i = g_thePlayerList->Get_Player_Count() - 1; i >= 0; i--) {
+                            Player *player = g_thePlayerList->Get_Nth_Player(i);
+
+                            if (Get_Controlling_Player()->Get_Relationship(player->Get_Default_Team()) == ALLIES) {
+                                index |= player->Get_Player_Mask();
+                            }
+                        }
+
+                        index2 = m_spiedOnPlayers | index;
+                    }
+
+                    Coord3D pos = *Get_Position();
+                    g_thePartitionManager->Do_Shroud_Reveal(pos.x, pos.y, radius, index2);
```

Comment:
`Do_Shroud_Reveal` should not be called last?

Reply:
This is correct.